### ISSUE TITLE
Timelock Server can issue decreasing and duplicated timestamps after network partition

### ISSUE THREAD

**Issue by jeremyk-91** [palantir/atlasdb#1310]
During @harrybiddle's Jepsen testing, we found that the Timelock Service could give duplicated and increasing timestamps. We probed further into the logs and found that this seemed to be an issue with recovery from network partitions. Consider the following logs:

```java
2016-11-29 13:20:19,846{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:invoke	:read-operation	nil
2016-11-29 13:20:19,847{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:invoke	:read-operation	nil
2016-11-29 13:20:19,864{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:ok	:read-operation	3681
2016-11-29 13:20:19,864{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:ok	:read-operation	3680
2016-11-29 13:20:20,748{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:start	"Cut off {:n3 #{:n4 :n2 :n5}, :n1 #{:n4 :n2 :n5}, :n4 #{:n3 :n1}, :n2 #{:n3 :n1}, :n5 #{:n3 :n1}}"
2016-11-29 13:20:43,527{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:fail	:read-operation	nil	feign.FeignException: status 500 reading TimestampService#getFreshTimestamp(); content:
{
  "message" : "a27c4c51-11d3-474d-8712-ec21a465a002",
  "exceptionClass" : "com.google.common.util.concurrent.UncheckedExecutionException",
  "stackTrace" : null
}
2016-11-29 13:20:40,749{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:stop	nil
2016-11-29 13:20:40,961{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:stop	"fully connected"
2016-11-29 13:20:43,570{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:invoke	:read-operation	nil
2016-11-29 13:20:43,582{GMT}	INFO	[jepsen worker 4] jepsen.util: 4	:ok	:read-operation	3680 // what??
```

This is wrong on two counts: it is a duplicate, and also does not obey the contract that timestamps are monotonically increasing. Looking at the other Jepsen messages, it seems that what happened was as follows:

* n3 was the leader
* At 13:20:20, the Jepsen nemesis created a network partition, splitting the network to {n1 n3} and {n2 n4 n5}.
* At 13:20:40, the Jepsen nemesis restored the network.

Looking at the logs for n3:

```java
10.0.42.1 - - [29/Nov/2016:13:20:19 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 3
10.0.42.1 - - [29/Nov/2016:13:20:19 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 6
WARN  [2016-11-29 13:20:20,640] io.atomix.copycat.server.state.LeaderAppender: n3/10.0.0.4:8700 - AppendRequest to n4/10.0.0.5:8700 failed. Reason: request timed out
(some comments about netty failing)
WARN  [2016-11-29 13:20:25,643] io.atomix.copycat.server.state.LeaderAppender: n3/10.0.0.4:8700 - Suspected network partition. Stepping down
INFO  [2016-11-29 13:20:25,654] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to FOLLOWER
(keep polling everyone)
ERROR [2016-11-29 13:20:43,485] com.palantir.remoting1.servers.jersey.JsonExceptionMapper: Error 3595d8f0-3290-4e39-bf6f-9620c4766b54: io.atomix.copycat.session.ClosedSessionException: session closed
ERROR [2016-11-29 13:20:43,493] com.palantir.remoting1.servers.jersey.JsonExceptionMapper: Error 6d995374-2b57-4e85-9241-9db2a41d5b37: Could not contact the cluster. Has this node (6083989b-fa2e-4284-bcf0-b466298ab7e8) been partitioned off?
10.0.42.1 - - [29/Nov/2016:13:20:20 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 500 165 "-" "okhttp/2.5.0" 23464
10.0.42.1 - - [29/Nov/2016:13:20:22 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 503 143 "-" "okhttp/2.5.0" 20508
10.0.42.1 - - [29/Nov/2016:13:20:38 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 503 143 "-" "okhttp/2.5.0" 5272
10.0.42.1 - - [29/Nov/2016:13:20:19 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 500 165 "-" "okhttp/2.5.0" 23537
10.0.42.1 - - [29/Nov/2016:13:20:25 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 503 143 "-" "okhttp/2.5.0" 17532
10.0.42.1 - - [29/Nov/2016:13:20:38 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 503 143 "-" "okhttp/2.5.0" 5280
(more of the "partitioned off?" errors, which are fine)
ERROR [2016-11-29 13:20:43,505] com.palantir.remoting1.servers.jersey.JsonExceptionMapper: Error 5bd6fdb8-f015-40a8-acae-ea4b57fae576: io.atomix.copycat.error.UnknownSessionException: unknown member session
ERROR [2016-11-29 13:20:43,509] com.palantir.remoting1.servers.jersey.JsonExceptionMapper: Error ee2f2753-b3bf-407b-9f87-a34e50e041c5: Could not contact the cluster. Has this node (6083989b-fa2e-4284-bcf0-b466298ab7e8) been partitioned off?
(more of the "partitioned off?" errors, which are unusual at this point!)
INFO  [2016-11-29 13:20:43,538] io.atomix.copycat.client.session.ClientSession: Registered session 3999
10.0.42.1 - - [29/Nov/2016:13:20:43 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 8
10.0.42.1 - - [29/Nov/2016:13:20:43 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 5
10.0.42.1 - - [29/Nov/2016:13:20:43 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 4
(gives out more timestamps)
INFO  [2016-11-29 13:20:47,330] io.atomix.catalyst.transport.netty.NettyClient: Connecting to n2/10.0.0.3:8700
INFO  [2016-11-29 13:20:47,701] io.atomix.copycat.server.state.FollowerState: n3/10.0.0.4:8700 - Polling members [ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n1/10.0.0.2:8700, clientAddress=n1/10.0.0.2:8700], ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n2/10.0.0.3:8700, clientAddress=n2/10.0.0.3:8700], ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n4/10.0.0.5:8700, clientAddress=n4/10.0.0.5:8700], ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n5/10.0.0.6:8700, clientAddress=n5/10.0.0.6:8700]]
INFO  [2016-11-29 13:20:47,702] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to CANDIDATE
INFO  [2016-11-29 13:20:47,702] io.atomix.copycat.server.state.CandidateState: n3/10.0.0.4:8700 - Starting election
INFO  [2016-11-29 13:20:47,731] io.atomix.copycat.server.state.CandidateState: n3/10.0.0.4:8700 - Requesting votes from [ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n1/10.0.0.2:8700, clientAddress=n1/10.0.0.2:8700], ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n2/10.0.0.3:8700, clientAddress=n2/10.0.0.3:8700], ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n4/10.0.0.5:8700, clientAddress=n4/10.0.0.5:8700], ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n5/10.0.0.6:8700, clientAddress=n5/10.0.0.6:8700]]
INFO  [2016-11-29 13:20:47,762] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to LEADER
```

Let's probe a bit further into the logs; during the partition, n2 transitioned to be the leader, but was never recognised as such (I guess it failed to write to the distributed LeaderAndTerm).

```java
INFO  [2016-11-29 13:20:23,218] io.atomix.copycat.server.state.ServerContext: n2/10.0.0.3:8700 - Transitioning to LEADER
INFO  [2016-11-29 13:20:23,221] io.atomix.copycat.server.state.ServerContext: n2/10.0.0.3:8700 - Found leader n2/10.0.0.3:8700
WARN  [2016-11-29 13:20:23,909] io.atomix.copycat.server.state.LeaderAppender: n2/10.0.0.3:8700 - Failed to configure n1/10.0.0.2:8700
WARN  [2016-11-29 13:20:23,910] io.atomix.copycat.server.state.LeaderAppender: n2/10.0.0.3:8700 - AppendRequest to n1/10.0.0.2:8700 failed. Reason: request timed out
WARN  [2016-11-29 13:20:23,953] io.atomix.copycat.server.state.LeaderAppender: n2/10.0.0.3:8700 - Failed to configure n3/10.0.0.4:8700
WARN  [2016-11-29 13:20:23,954] io.atomix.copycat.server.state.LeaderAppender: n2/10.0.0.3:8700 - AppendRequest to n3/10.0.0.4:8700 failed. Reason: request timed out
INFO  [2016-11-29 13:20:23,987] io.atomix.catalyst.transport.netty.NettyClient: Connecting to n1/10.0.0.2:8700
INFO  [2016-11-29 13:20:23,987] io.atomix.catalyst.transport.netty.NettyClient: Connecting to n3/10.0.0.4:8700
INFO  [2016-11-29 13:20:24,238] io.atomix.catalyst.transport.netty.NettyClient: Connecting to n1/10.0.0.2:8700
INFO  [2016-11-29 13:20:24,238] io.atomix.catalyst.transport.netty.NettyClient: Connecting to n3/10.0.0.4:8700
INFO  [2016-11-29 13:20:26,103] io.atomix.catalyst.transport.netty.NettyClient: Connecting to n5/10.0.0.6:8700
INFO  [2016-11-29 13:20:26,104] io.atomix.catalyst.transport.netty.NettyClient: Connected to n5/10.0.0.6:8700
WARN  [2016-11-29 13:20:28,992] io.atomix.copycat.server.state.LeaderAppender: n2/10.0.0.3:8700 - AppendRequest to n1/10.0.0.2:8700 failed. Reason: io.netty.channel.ConnectTimeoutException: connection timed out: n1/10.0.0.2:8700
WARN  [2016-11-29 13:20:28,992] io.atomix.copycat.server.state.LeaderAppender: n2/10.0.0.3:8700 - AppendRequest to n3/10.0.0.4:8700 failed. Reason: io.netty.channel.ConnectTimeoutException: connection timed out: n3/10.0.0.4:8700
INFO  [2016-11-29 13:20:28,992] io.atomix.catalyst.transport.netty.NettyClient: Connecting to n1/10.0.0.2:8700
INFO  [2016-11-29 13:20:28,998] io.atomix.catalyst.transport.netty.NettyClient: Connecting to n3/10.0.0.4:8700
ERROR [2016-11-29 13:20:29,004] com.palantir.remoting1.servers.jersey.JsonExceptionMapper: Error 47c6e030-66e7-48d4-a066-1551ddb796ec: This node (d2566ef2-d33d-4270-b79f-cd79dad52b93) is not the leader (LeaderAndTerm{term=8, leaderId=6083989b-fa2e-4284-bcf0-b466298ab7e8})
```

I'm in the process of investigating this further.

Filing as P0, as this is a correctness bug (and happened with just a few thousand timestamps!)

**Comment by jeremyk-91** [comment]
Some investigation: `LeaderAndTerm` and the actual distributed group leader seem like they can get out of sync.

```java
jkong-mbp:latest jkong$ for i in n1 n2 n3 n4 n5; do cat "$i/atlasdb-timelock-server-startup.log" | grep "Transitioning to"; done
INFO  [2016-11-29 13:18:23,514] io.atomix.copycat.server.state.ServerContext: n1/10.0.0.2:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:19:03,997] io.atomix.copycat.server.state.ServerContext: n1/10.0.0.2:8700 - Transitioning to CANDIDATE
INFO  [2016-11-29 13:19:04,144] io.atomix.copycat.server.state.ServerContext: n1/10.0.0.2:8700 - Transitioning to LEADER
INFO  [2016-11-29 13:19:32,906] io.atomix.copycat.server.state.ServerContext: n1/10.0.0.2:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:21:16,294] io.atomix.copycat.server.state.ServerContext: n1/10.0.0.2:8700 - Transitioning to CANDIDATE
INFO  [2016-11-29 13:21:16,353] io.atomix.copycat.server.state.ServerContext: n1/10.0.0.2:8700 - Transitioning to LEADER
INFO  [2016-11-29 13:21:44,411] io.atomix.copycat.server.state.ServerContext: n1/10.0.0.2:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:22:00,671] io.atomix.copycat.server.state.ServerContext: n1/10.0.0.2:8700 - Transitioning to CANDIDATE
INFO  [2016-11-29 13:22:00,734] io.atomix.copycat.server.state.ServerContext: n1/10.0.0.2:8700 - Transitioning to LEADER
INFO  [2016-11-29 13:22:37,491] io.atomix.copycat.server.state.ServerContext: n1/10.0.0.2:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:18:20,997] io.atomix.copycat.server.state.ServerContext: n2/10.0.0.3:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:18:24,250] io.atomix.copycat.server.state.ServerContext: n2/10.0.0.3:8700 - Transitioning to CANDIDATE
INFO  [2016-11-29 13:18:24,333] io.atomix.copycat.server.state.ServerContext: n2/10.0.0.3:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:20:23,149] io.atomix.copycat.server.state.ServerContext: n2/10.0.0.3:8700 - Transitioning to CANDIDATE
INFO  [2016-11-29 13:20:23,218] io.atomix.copycat.server.state.ServerContext: n2/10.0.0.3:8700 - Transitioning to LEADER
INFO  [2016-11-29 13:20:52,005] io.atomix.copycat.server.state.ServerContext: n2/10.0.0.3:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:18:23,408] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:18:24,211] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to CANDIDATE
INFO  [2016-11-29 13:18:24,328] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to LEADER
INFO  [2016-11-29 13:19:06,073] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:19:28,341] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to CANDIDATE
INFO  [2016-11-29 13:19:28,402] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to LEADER
INFO  [2016-11-29 13:20:25,654] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:20:47,702] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to CANDIDATE
INFO  [2016-11-29 13:20:47,762] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to LEADER
INFO  [2016-11-29 13:21:18,345] io.atomix.copycat.server.state.ServerContext: n3/10.0.0.4:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:18:23,366] io.atomix.copycat.server.state.ServerContext: n4/10.0.0.5:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:18:23,522] io.atomix.copycat.server.state.ServerContext: n5/10.0.0.6:8700 - Transitioning to FOLLOWER
INFO  [2016-11-29 13:22:39,441] io.atomix.copycat.server.state.ServerContext: n5/10.0.0.6:8700 - Transitioning to CANDIDATE
INFO  [2016-11-29 13:22:39,513] io.atomix.copycat.server.state.ServerContext: n5/10.0.0.6:8700 - Transitioning to LEADER
jkong-mbp:latest jkong$ for i in n1 n2 n3 n4 n5; do cat "$i/atlasdb-timelock-server-startup.log" | grep "Leader has been set to"; done
INFO  [2016-11-29 13:18:24,735] com.palantir.atlasdb.timelock.TimeLockServer: Leader has been set to LeaderAndTerm{term=8, leaderId=6083989b-fa2e-4284-bcf0-b466298ab7e8}
jkong-mbp:latest jkong$ for i in n1 n2 n3 n4 n5; do cat "$i/atlasdb-timelock-server-startup.log" | grep "Not setting the leader to"; done
jkong-mbp:latest jkong$ for i in n1 n2 n3 n4 n5; do cat "$i/atlasdb-timelock-server-startup.log" | grep "Updating the leader"; done
```

I'm trying to match this up with the messages we may have had in the logs about ClosedSession and UnknownSessionExceptions - are we failing the distributed writes for some reason, and if so would that cause the `LeaderAndTerm` to become out of sync with the actual leader election?

**Comment by harrybiddle** [comment]
The other interesting thing in this run is that during the network partition that you refer to, the _timelock server did not give out any timestamps at all_. Instead, it gave out some 500's that aborted the round-robining of the client. In general, I get a lot of `ClosedSession` and `UnknownSessionExceptions` that cause 500's in the Jepsen runs. I'd like for the Jepsen tests to never catch a fail at all: our server should be totally robust to partitioning the network, is that right?

**Comment by jeremyk-91** [comment]
@harrybiddle 

Yes. In a two-way partition with the leader in the minority, the expected behaviour is for:
* the leader to figure out it is not the leader and step down (this eventually did happen)
* the majority partition to elect a new leader (this kind of happened, but not quite; previous post shows that during the partition an election did happen and n2 did become the leader, but `LeaderAndTerm` was not updated).

`ClosedSessionException` is addressed by PR #1291, I'm trying to figure out but still unsure as to what's causing the `UnknownSessionException`s.

**Comment by jeremyk-91** [comment]
Also, specifically regarding your comment: `LeaderAndTerm` is not being updated properly, so actually no one other than `n3` was able to give out timestamps, even if they might have been the leader. Furthermore, even if `n3` was not the leader, it was always able to give out timestamps.

**Comment by SerialVelocity** [comment]
`UnknownSessionException` is caused by bad session recovery. This has been fixed `>= 1.0.0-rc9` but no release has been cut yet.

**Comment by SerialVelocity** [comment]
`LeaderAndTerm` not being equal to the current leader doesn't matter since we don't actually care who the leader is. We just need for there to only be one server giving out locks.

**Comment by jeremyk-91** [comment]
I think I've got a bit more insight into the problem, after some debugging.

Firstly, there are multiple leaders in place when we run Atomix:
* When we create an Atomix cluster, we use Raft to decide a group leader.
* The `DistributedGroup` is a separate notion, which sits on top of the Atomix cluster; its leader election is distinct from the leader election of the Atomix cluster.

Consider that all Timelock operations are dependent on the `DistributedGroup` leader, not the cluster leader. This is because that's what's used to write to `LeaderAndTerm`. Furthermore, if a server leaves the network and rejoins it owing to a closed session, a new client is constructed, with a new session ID. Critically, _this means it is no longer a member of the `DistributedGroup`_. The other nodes will do a re-election on the reduced group.

In fact, this is consistent with what happened during the Jepsen run. Investigating the behaviour of the nemesis, we had

```java
2016-11-29 13:18:33,806{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:start	nil
2016-11-29 13:18:34,985{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:start	"Cut off {:n2 #{:n3 :n4 :n5}, :n1 #{:n3 :n4 :n5}, :n3 #{:n2 :n1}, :n4 #{:n2 :n1}, :n5 #{:n2 :n1}}"
2016-11-29 13:18:54,987{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:stop	nil
2016-11-29 13:18:55,198{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:stop	"fully connected"
2016-11-29 13:19:00,200{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:start	nil
2016-11-29 13:19:01,380{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:start	"Cut off {:n3 #{:n4 :n2 :n1}, :n5 #{:n4 :n2 :n1}, :n4 #{:n3 :n5}, :n2 #{:n3 :n5}, :n1 #{:n3 :n5}}" // no news about a re-election?
2016-11-29 13:19:21,381{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:stop	nil
2016-11-29 13:19:21,593{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:stop	"fully connected"
2016-11-29 13:19:26,594{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:start	nil
2016-11-29 13:19:27,935{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:start	"Cut off {:n5 #{:n3 :n4 :n2}, :n1 #{:n3 :n4 :n2}, :n3 #{:n5 :n1}, :n4 #{:n5 :n1}, :n2 #{:n5 :n1}}"
2016-11-29 13:19:47,936{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:stop	nil
2016-11-29 13:19:48,145{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:stop	"fully connected"
2016-11-29 13:19:53,146{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:start	nil
2016-11-29 13:19:54,408{GMT}	INFO	[jepsen nemesis] jepsen.util: :nemesis	:info	:start	"Cut off {:n4 #{:n3 :n5 :n1}, :n2 #{:n3 :n5 :n1}, :n3 #{:n4 :n2}, :n5 #{:n4 :n2}, :n1 #{:n4 :n2}}"
```

Observe that during the first partition the majority is {n3, n4, n5} including the then-leader. From the cluster's point of view, n1 and n2 left. So the `timeLockGroup` likely still had n3, n4 and n5; since n3 was the starting leader, no re-election took place. The next group would have had n4 leave, and the one after that have n5 leave, leaving n3 as the _only_ leader of the `timeLockGroup` regardless of whatever happens in the Atomix groups.

Running more tests with @harrybiddle that feature additional logging now, but offhand it seems like a reasonable solution could be to use persistent groups (but we need to explore the semantics of that).

Some of this could be fixed by atomix/atomix#190. We might want to see if there's a way we can work around this, as well.

**Comment by jeremyk-91** [comment]
Verified that so far, all failures seen in tests have occurred immediately after recovery from a network partition in which our leader was not in the majority. The last few writes don't seem to be "remembered", in that sense. This terminates the session, and subsequently a new session is started when we recover (the failures seem to be that the last few writes don't make it through).

**Comment by jeremyk-91** [comment]
Further updates:
* Apparently, it is possible for the local state (by reading the distributed long with `ReadConsistency.SEQUENTIAL`) to get ahead on some nodes (say, to `8715`), and then "snap back" to a lower value (e.g. `8711`) after the partition is actually resolved. Harry and I have turned on debug logging to nail down what's going on.

**Comment by jeremyk-91** [comment]
More updates, after adding more logging to the Jepsen tests. We had something like this on one of the boxes - I added an additional thread that would repeatedly sleep for 0.05s and then attempt to read the distributed long with sequential consistency (this gives us the local value, because we don't provide a bound on the currency of the data) and log it.

```java
INFO  [2016-12-01 18:54:04,249] com.palantir.atlasdb.timelock.TimeLockServer: I believe the long is 7072
10.0.42.1 - - [01/Dec/2016:18:54:04 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 47
10.0.42.1 - - [01/Dec/2016:18:54:04 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 60
10.0.42.1 - - [01/Dec/2016:18:54:04 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 39
INFO  [2016-12-01 18:54:04,332] com.palantir.atlasdb.timelock.TimeLockServer: I believe the long is 7075
10.0.42.1 - - [01/Dec/2016:18:54:04 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 6
10.0.42.1 - - [01/Dec/2016:18:54:04 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 14
INFO  [2016-12-01 18:54:04,391] com.palantir.atlasdb.timelock.TimeLockServer: I believe the long is 7077
10.0.42.1 - - [01/Dec/2016:18:54:04 +0000] "POST /test/timestamp/fresh-timestamp HTTP/1.1" 200 4 "-" "okhttp/2.5.0" 36
INFO  [2016-12-01 18:54:05,136] io.atomix.copycat.server.state.FollowerState: n5/10.0.0.6:8700 - Polling members [ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n1/10.0.0.2:8700, clientAddress=n1/10.0.0.2:8700], ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n2/10.0.0.3:8700, clientAddress=n2/10.0.0.3:8700], ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n3/10.0.0.4:8700, clientAddress=n3/10.0.0.4:8700], ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=n4/10.0.0.5:8700, clientAddress=n4/10.0.0.5:8700]]
(... some other stuff about not being able to connect etc.)
INFO  [2016-12-01 18:54:13,745] io.atomix.copycat.client.session.ClientSession: Registered session 7675
INFO  [2016-12-01 18:54:13,748] com.palantir.atlasdb.timelock.TimeLockServer: I believe the long is 7074
INFO  [2016-12-01 18:54:13,801] com.palantir.atlasdb.timelock.TimeLockServer: I believe the long is 7074
```

My suspicion is that some updates are being made perhaps when the client is in an `UNSTABLE` state, but I still don't have a concrete answer. Nonetheless we are now pretty sure that single clients can have inconsistent views across sessions - we need to figure out why this is happening.

After the partition is lifted, everyone starts a new session where they believe the long has slid back to 7074 (despite locally having views of 7078 or 7079, apart from n1 which curiously had only 7065). I think this might have to do with the `DistributedLong`'s state - will add some logging tomorrow!

At this point I'm tempted to fix some of the other bugs which we are aware of, which may not be P0 but are making the debugging more difficult to reason about, like #1315 and #1320.

**Comment by jeremyk-91** [comment]
An alternative could be to try and base leadership decisions based on whether the timestamp has changed out from under us.
